pkg/scheduler/plugins/elasticquota/core/group_quota_manager.go

```diff
@@ -0,0 +1,616 @@
+// of all quotaInfo is done by gqm. scopedLockForQuotaInfo, so just get treeWrappers' lock when calling treeWrappers' function
+func (gqm *GroupQuotaManager) updateGroupDeltaRequestTopoRecursiveNoLock(deltaReq v1.ResourceList, curToAllParInfos []*QuotaInfo) {
+	quotaInfo := curToAllParInfos[0]
+	oldSubLimitReq := quotaInfo.getLimitRequestNoLock()
+	quotaInfo.addRequestNonNegativeNoLock(deltaReq)
+	newSubLimitReq := quotaInfo.getLimitRequestNoLock()
+	treeWrapperPtr := gqm.getTreeWrapperByNameNoLock(quotaInfo.ParentName)
+	if treeWrapperPtr == nil {
+		klog.Errorf("treeWrapper not exist! quotaInfoName:%v parentName:%s", quotaInfo.Name, quotaInfo.ParentName)
+		return
+	}
+	if treeWrapperPtr.NeedUpdateOneGroupRequest(quotaInfo) {
+		treeWrapperPtr.UpdateOneGroupRequest(quotaInfo)
+	}
+
+	for i := 1; i < len(curToAllParInfos); i++ {
+		parentQuotaInfo := curToAllParInfos[i]
+		directParTreeWrapperPtr := gqm.getTreeWrapperByNameNoLock(parentQuotaInfo.ParentName)
+		if directParTreeWrapperPtr == nil {
+			klog.Errorf("treeWrapper not exist!  parentName:%v", quotaInfo.Name, quotaInfo.ParentName)
+			return
+		}
+
+		oldSubLimitReqTmp := quotav1.Subtract(v1.ResourceList{}, oldSubLimitReq)
+		oldSubLimitReq = parentQuotaInfo.getLimitRequestNoLock()
+		parentQuotaInfo.addRequestNonNegativeNoLock(oldSubLimitReqTmp)
+		parentQuotaInfo.addRequestNonNegativeNoLock(newSubLimitReq)
+		newSubLimitReq = parentQuotaInfo.getLimitRequestNoLock()
+
+		if directParTreeWrapperPtr.NeedUpdateOneGroupRequest(parentQuotaInfo) {
+			directParTreeWrapperPtr.UpdateOneGroupRequest(parentQuotaInfo)
+		}
+	}
+}
+
+func (gqm *GroupQuotaManager) UpdateGroupDeltaUsed(quotaName string, delta v1.ResourceList) {
+	gqm.Lock.RLock()
+	defer gqm.Lock.RUnlock()
+
+	gqm.updateGroupDeltaUsedNoLock(quotaName, delta)
+	// if systemQuotaGroup or DefaultQuotaGroup's used change, update cluster total resource.
+	if quotaName == extension.SystemQuotaName || quotaName == extension.DefaultQuotaName {
+		gqm.updateClusterTotalResourceNoLock(v1.ResourceList{})
+	}
+}
+
+// updateGroupDeltaUsedNoLock updates the usedQuota of a node, it also updates all parent nodes
+// no need to lock gqm.lock
+func (gqm *GroupQuotaManager) updateGroupDeltaUsedNoLock(quotaName string, delta v1.ResourceList) {
+	curToAllParInfos := gqm.getCurToAllParentGroupQuotaInfoNoLock(quotaName)
+	allQuotaInfoLen := len(curToAllParInfos)
+	if allQuotaInfoLen <= 0 {
+		return
+	}
+
+	defer gqm.scopedLockForQuotaInfo(curToAllParInfos)()
+	for i := 0; i < allQuotaInfoLen; i++ {
+		quotaInfo := curToAllParInfos[i]
+		quotaInfo.addUsedNonNegativeNoLock(delta)
+	}
+}
+
+func (gqm *GroupQuotaManager) RefreshRuntime(quotaName string) v1.ResourceList {
+	gqm.Lock.RLock()
+	defer gqm.Lock.RUnlock()
+
+	return gqm.refreshRuntimeNoLock(quotaName)
+}
+
+func (gqm *GroupQuotaManager) refreshRuntimeNoLock(quotaName string) v1.ResourceList {
```

Comment:
refreshRuntimeNoLock and refreshAndGetRuntimeQuotaNoLock seems a little bit dup, keep one is enough